omnipaxos_storage/src/persistent_storage.rs

```diff
@@ -0,0 +1,384 @@
+    }
+
+    pub fn set_rocksdb_path(&mut self, path: String) {
+        self.rocksdb_path = Some(path);
+    }
+
+    pub fn get_commitlog_options(&self) -> LogOptions {
+        self.commitlog_options.clone()
+    }
+
+    pub fn set_commitlog_options(&mut self, commitlog_opts: LogOptions) {
+        self.commitlog_options = commitlog_opts;
+    }
+
+    pub fn get_rocksdb_options(&self) -> Options {
+        self.rocksdb_options.clone()
+    }
+
+    pub fn set_rocksdb_options(&mut self, rocksdb_opts: Options) {
+        self.rocksdb_options = rocksdb_opts;
+    }
+
+    pub fn with(
+        log_path: String,
+        commitlog_options: LogOptions,
+        rocksdb_path: String,
+        rocksdb_options: Options,
+    ) -> Self {
+        Self {
+            commitlog_path: Some(log_path),
+            commitlog_options,
+            rocksdb_path: Some(rocksdb_path),
+            rocksdb_options,
+        }
+    }
+}
+
+impl Default for PersistentStorageConfig {
+    fn default() -> Self {
+        let mut rocksdb_options = Options::default();
+        rocksdb_options.create_if_missing(true); // Creates rocksdb store if it is missing
+
+        Self {
+            commitlog_path: Some(COMMITLOG.to_string()),
+            commitlog_options: LogOptions::new(COMMITLOG),
+            rocksdb_path: Some(ROCKSDB.to_string()),
+            rocksdb_options,
+        }
+    }
+}
+
+/// A persistent storage implementation, lets sequence paxos write the log
+/// and variables (e.g. promised, accepted) to disk. Log entries are serialized
+/// and de-serialized into slice of bytes when read or written from the log.
+pub struct PersistentStorage<T, S>
+where
+    T: Entry,
+    S: Snapshot<T>,
+{
+    /// disk-based commit log for entries
+    commitlog: CommitLog,
+    /// the path to the directory containing a commitlog
+    log_path: String,
+    /// local RocksDB key-value store
+    rocksdb: DB,
+    /// A placeholder for the T: Entry
+    t: PhantomData<T>,
+    /// A placeholder for the S: Snapshot<T>
+    s: PhantomData<S>,
+}
+
+impl<T: Entry, S: Snapshot<T>> PersistentStorage<T, S> {
+    pub fn with(storage_config: PersistentStorageConfig) -> Self {
+        // Paths to commitlog and rocksDB store
+        let log_path: String = storage_config.commitlog_path.unwrap();
+        let rocks_path = storage_config.rocksdb_path.unwrap();
+
+        // Check if storage already exists on the paths
+        std::fs::metadata(&log_path)
+            .expect_err(&format!("commitlog already exists in {}", log_path));
+        std::fs::metadata(&rocks_path)
+            .expect_err(&format!("rocksDB store already exists in {}", rocks_path));
+
+        // Initialize Commitlog and rocksDB
+        let commitlog =
+            CommitLog::new(storage_config.commitlog_options).expect("Failed to create Commitlog");
+        let rocksdb = DB::open(&storage_config.rocksdb_options, &rocks_path)
+            .expect("Failed to create rocksDB store");
```

Comment:
For rocks I think you could use this [option](https://docs.rs/rocksdb/latest/rocksdb/struct.Options.html#method.set_error_if_exists) and remove the check in the beginning. 

Furthermore, I'm not sure we need the check. Won't commitlog and rocks panic anyway if we try to create something on a path is already in use?

Reply:
I agree the check is no longer needed, as both Commitlog and rocksDB will either create a new storage or open existing ones in a given path.